libs/datalayout/src/main/java/com/linkedin/openhouse/datalayout/generator/OpenHouseDataLayoutGenerator.java

```diff
@@ -0,0 +1,113 @@
+package com.linkedin.openhouse.datalayout.generator;
+
+import com.linkedin.openhouse.datalayout.config.DataCompactionConfig;
+import com.linkedin.openhouse.datalayout.datasource.FileStat;
+import com.linkedin.openhouse.datalayout.datasource.TableFileStats;
+import com.linkedin.openhouse.datalayout.datasource.TablePartitionStats;
+import com.linkedin.openhouse.datalayout.strategy.DataLayoutOptimizationStrategy;
+import java.util.Collections;
+import java.util.List;
+import lombok.Builder;
+import org.apache.spark.api.java.function.FilterFunction;
+import org.apache.spark.api.java.function.MapFunction;
+import org.apache.spark.sql.Encoders;
+
+/**
+ * Data layout optimization strategies generator for OpenHouse. Generates a list of strategies with
+ * cost.
+ */
+@Builder
+public class OpenHouseDataLayoutGenerator implements DataLayoutGenerator {
+  private static final long MB = 1024 * 1024;
+  private static final long FILE_BLOCK_SIZE_BYTES = 256 * MB;
+  private static final long FILE_BLOCK_MARGIN_BYTES = 10 * MB;
+  private static final long NUM_FILE_GROUPS_PER_COMMIT = 100;
+  private static final long MAX_NUM_COMMITS = 30;
+  private static final long MAX_BYTES_SIZE_RATIO = 10;
+  private static final long REWRITE_BYTES_PER_SECOND = 2 * MB;
+  private static final long EXECUTOR_MEMORY_MB = 2048;
+  private static final int MAX_CONCURRENT_FILE_GROUP_REWRITES = 50;
+  private static final int REWRITE_PARALLELISM = 900; // number of Spark tasks to run in parallel
+  private static final long TARGET_BYTES_SIZE = 2 * FILE_BLOCK_SIZE_BYTES - FILE_BLOCK_MARGIN_BYTES;
+  private static final double COMPUTE_STARTUP_COST_GB_HR = 0.5;
+  private final TableFileStats tableFileStats;
+  private final TablePartitionStats tablePartitionStats;
+
+  /**
+   * Generate a list of data layout optimization strategies based on the table file stats and
+   * historic query patterns.
+   */
+  @Override
+  public List<DataLayoutOptimizationStrategy> generate() {
+    // skip single partition and non-partitioned tables
+    if (tablePartitionStats.get().count() <= 1) {
+      return Collections.emptyList();
+    }
+    return Collections.singletonList(generateCompactionStrategy());
+  }
+
+  private DataLayoutOptimizationStrategy generateCompactionStrategy() {
```

Comment:
```suggestion
  /**
   * Method to compute the file right sizing strategy for a partitioned table. In this method we compute number of files that are sized lesser than a threshold, compute predicted cost and predicted gain and finally computing a score for the strategy.
   * predicted cost is computed as the GBhr spent in performing the compaction.
   * predicted gain is computed as the file counts reduced.
   * predicted score is computed as gain / cost, the number of files reduced per GB-hr of compute. Higher the score, better is the strategy.
   */
  private DataLayoutOptimizationStrategy computePartitionedTableFileRightSizingStrategy() {
```

@anjagruenheid this is what Stas decided for cost, gain and score for file rightsizing. We will have a to manually iterate on threshold for score for picking a table for compaction. Further we will compare the predicted cost, gain, and score with actual cost, gain, and score and check how close they are. Please let us know what you think.

Reply:
IMO, compaction is a very broad term so I was suggesting something very specific. One could argue that sorting improves compressions and hence compacts data.

Reply:
Compaction is Iceberg terminology for this operation https://iceberg.apache.org/docs/latest/maintenance/#compact-data-files, it become more like compaction when we consider delete files. Right-sizing has resource allocation relation, although it's more specific.

Reply:
Fair. Let's do compaction